lib/Model/Pass/DeduplicateCollidingNames.cpp

```diff
@@ -0,0 +1,185 @@
+/// \file DeduplicateCollidingNames.cpp
+/// Implementation of deduplication of colliding names.
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <unordered_set>
+
+#include "revng/Model/Binary.h"
+#include "revng/Model/NameBuilder.h"
+#include "revng/Model/Pass/DeduplicateCollidingNames.h"
+#include "revng/Model/Pass/RegisterModelPass.h"
+
+static RegisterModelPass R("deduplicate-colliding-names",
+                           "This looks for groups of names are duplicates of "
+                           "each other and appends arbitrary suffixes "
+                           "to prevent collisions.",
+                           model::deduplicateCollidingNames);
+
+struct Namespace {
+public:
+  using NameSet = std::unordered_set<std::string,
+                                     model::detail::TransparentStringHash>;
+
+  struct CollisionList {
+    std::vector<std::string *> Names = {};
+    size_t FinalOffset = 0;
+  };
+  using CollisionMap = std::map<std::string, CollisionList>;
+
+public:
+  const model::NamingConfiguration &Configuration;
+  const Namespace *ParentNamespace = nullptr;
+
+  NameSet Names = {};
+  CollisionMap Collisions = {};
+
+public:
+  Namespace(const model::NamingConfiguration &Configuration,
+            const Namespace &ParentNamespace) :
+    Configuration(Configuration), ParentNamespace(&ParentNamespace) {}
+  Namespace(const model::NamingConfiguration &Configuration) :
+    Configuration(Configuration), ParentNamespace(nullptr) {}
+
+public:
+  void add(std::string &Name) {
+    if (not ParentNamespace or not ParentNamespace->Names.contains(Name)) {
+      auto [Iterator, Success] = Names.emplace(Name);
+      if (Success) {
+        // This is the first time we see this name, nothing else to do.
+        return;
+      }
+    }
+
+    // Mark this as a collision.
+    Collisions[Name].Names.emplace_back(&Name);
+  }
+
+  void resolve() {
+    for (auto &&[SanitizedName, CollisionList] : Collisions) {
+      revng_assert(CollisionList.FinalOffset <= CollisionList.Names.size(),
+                   "`resolve` can only be called once per namespace.");
+
+      uint64_t SuffixIndex = 0;
+      if (ParentNamespace) {
+        auto Iterator = ParentNamespace->Collisions.find(SanitizedName);
+        if (Iterator != ParentNamespace->Collisions.end())
+          SuffixIndex += Iterator->second.FinalOffset;
+      }
+
+      static constexpr llvm::StringRef CollisionResolutionSuffix = "_";
+      for (std::string *Entry : CollisionList.Names) {
+        auto R = SanitizedName + CollisionResolutionSuffix.str()
+                 + std::to_string(++SuffixIndex);
+
+        while (Names.contains(R)
+               or (ParentNamespace and ParentNamespace->Names.contains(R))) {
+          R = SanitizedName + CollisionResolutionSuffix.str()
+              + std::to_string(++SuffixIndex);
```

Comment:
@aleclearmind, almost forgot, we still need to decide on the deduplication suffix.

Options:
1. What we had before:
	- `name`
	- `name_`
	- `name__`
	- `name___`
2. Current (indexes):
	- `name`
	- `name_1`
	- `name_2`
	- `name_3`
3. Mixed?:
	- `name`
	- `name_`
	- `name_2`
	- `name_3`
4. Something else?

P. S. I just realized that this still uses `SanitizedName` identifier, that's just a local name, I'll switch it to just `Name`.

Reply:
2) Looks good.

Maybe let's start from `_2` so we don't have `name` and `name_1` but `name` and `name_2`.
What do you think?